use existing retry logic for maven metadata download

hdeadman opened the pull request:
Fixes  #27262

This changes the existing retry logic to use a Callable instead of a Runnable so there can be a return value. Then it uses the retry logic for the maven metadata download.

ov7a commented on platforms/software/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publisher/DependencyResolverIvyPublisher.java:
```diff
@@ -43,10 +45,11 @@ public void publish(IvyNormalizedPublication publication, IvyArtifactRepository
+        networkOperationBackOffAndRetry.withBackoffAndRetry(new Callable<String>() {
             @Override
-            public void run() {
+            public String call() {
```
❌ Please use `Void` instead of `String` here.

ov7a commented on platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java:
```diff
@@ -273,10 +286,11 @@ private byte[] createChecksumFile(File src, HashFunction hashFunction) {
+            networkOperationCaller.withBackoffAndRetry(new Callable<String>() {
                 @Override
-                public void run() {
+                public String call() {
```
❌ Please use `Void` instead of `String` here.

ov7a reviewed (changes requested):
Thank you for your contribution!

This PR is complete and simple enough.

***

While it would be nice, I don't think we need tests here: the retry method is covered by its own test, and the publishing process is covered by a functional test.

@hdeadman Please address the comments about the types below and fix the failing tests.

hdeadman commented:
I fixed the return types to be Void instead of String. The failing test I saw said that internal API had to have the @NonNullApi annotation so I added the annotation to `AbstractMavenPublisher` and it's sub-classes, and adjusted some code based on that contract. I will try to run all the tests locally to see if there are more failing tests.

hdeadman commented:
Looks like I need to fix the tests, lots of failing maven tests when I run locally so I will work on those.

hdeadman commented:
`./gradlew :maven:quicktest -Dorg.gradle.ignoreBuildJavaVersionCheck=true` passed locally, and so did `InternalNullabilityTest` which failed before.

hdeadman commented:
When I would run `./gradlew architecture-test:test -ParchunitRefreeze` it modified all the files in changes\archunit_store by adding a line feed at the end of the file, so I didn't notice that is was adding a few lines to one of the files. I just pushed that change so hopefully the `InternalNullabilityTest` will pass in CI?

ov7a commented on platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/MavenRemotePublisher.java:
```diff
@@ -49,7 +52,8 @@ public MavenRemotePublisher(Factory<File> temporaryDirFactory, BuildCommencedTim
 
     @Override
-    public void publish(MavenNormalizedPublication publication, MavenArtifactRepository artifactRepository) {
+    public void publish(MavenNormalizedPublication publication, @Nullable MavenArtifactRepository artifactRepository) {
```
🤔 That's weird. Why did you declare `artifactRepository` as Nullable to have an assert it being not null on the line below?

ov7a replied:
(sorry for the late reply)

hdeadman replied:
I had to add the `@NonNullApi` annotation at the class level because one of the CI tests (InternalNullabilityTest?) was failing if I didn't. Since the base `MavenPublisher` interface declares that parameter as `@Nullable`, I had to say it was `@Nullable` in order to make Intellij warning go away. I added the assertion because in this particular implementation of the `MavenPublisher` interface, the parameter cannot be null. Should I remove assert and let it NPE on the next line?

![image](https://github.com/gradle/gradle/assets/1556777/bc854e39-f106-4a9f-8cac-aeb93290c824)

ov7a replied:
A-ha. Indeed, there is a problem with class hierarchy in that area. I don't think this should prevent this PR from merging. That being said, I think it's ok to merge this.

Thanks @hdeadman!

ov7a commented:
> When I would run ./gradlew architecture-test:test -ParchunitRefreeze it modified all the files in changes\archunit_store by adding a line feed at the end of the file, so I didn't notice that is was adding a few lines to one of the files. I just pushed that change so hopefully the InternalNullabilityTest will pass in CI?

Extra empty lines are okay. These text files contain a list of exceptions - if you don't add any new exceptions, it's ok.

ov7a commented:
@bot-gradle merge

bot-gradle commented:
I've triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=76051671) for you. Click [here](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.names=gitCommitId&amp;search.tags=CI&amp;search.values=7c2d837def90172c4c00bca8627e884bb33be446) to see all failures if there's any.
